src/ray/raylet/scheduling_policy.cc

```diff
@@ -116,36 +118,54 @@ std::unordered_map<TaskID, ClientID> SchedulingPolicy::Schedule(
   return decision;
 }
 
-std::vector<TaskID> SchedulingPolicy::SpillOver(
-    SchedulingResources &remote_scheduling_resources) const {
+std::unordered_map<TaskID, ClientID> SchedulingPolicy::SpillOver(
+    std::unordered_map<ClientID, SchedulingResources> &cluster_resources,
+    std::vector<ClientID> &remote_client_ids) {
+  // Shuffle the nodes to prevent infinite forwarding loops.
+  const unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+  auto rng = std::default_random_engine(seed);
+  std::shuffle(remote_client_ids.begin(), remote_client_ids.end(), rng);
   // The policy decision to be returned.
-  std::vector<TaskID> decision;
-
-  ResourceSet new_load(remote_scheduling_resources.GetLoadResources());
+  std::unordered_map<TaskID, ClientID> decision;
 
-  // Check if we can accommodate an infeasible task.
+  // Check if we can accommodate all infeasible tasks.
   for (const auto &task : scheduling_queue_.GetInfeasibleTasks()) {
     const auto &spec = task.GetTaskSpecification();
-    if (spec.GetRequiredPlacementResources().IsSubset(
-            remote_scheduling_resources.GetTotalResources())) {
-      decision.push_back(spec.TaskId());
-      new_load.AddResources(spec.GetRequiredResources());
-    }
-  }
-
-  for (const auto &task : scheduling_queue_.GetReadyTasks()) {
-    const auto &spec = task.GetTaskSpecification();
-    if (!spec.IsActorTask()) {
-      if (spec.GetRequiredPlacementResources().IsSubset(
-              remote_scheduling_resources.GetTotalResources())) {
-        decision.push_back(spec.TaskId());
+    const auto &placement_resources = spec.GetRequiredPlacementResources();
+    // Find the first node that can accommodate the resources.
+    // NOTE(ujvl): this does not distribute tasks evenly across nodes.
+    for (auto &client_id : remote_client_ids) {
```

Comment:
The random shuffle of remote client ids above doesn't address the problem. In this for loop you are iterating over the remote client ids in the same order for all infeasible tasks. The result of this is still sending all infeasible tasks to the first remote client id in the shuffled list.

Reply:
Yeah you're right, I initially toyed with generating indices uniformly at random to distribute the load more uniformly, but it was getting too complicated: for example, if you try to spill over to a random node but it doesn't have the set of resources required, we'd have to try again (or give up) on it etc. 

Figured that shuffling would be a quick fix to enforce that between heartbeats things are random so any given task is less likely to be forwarded in cycles. But yes, over a single heartbeat it's going to end up dumping everything on one node.